Foundation/CPCache.j

```diff
@@ -0,0 +1,350 @@
+    @param aCost the object's cost
+*/
+- (void)setObject:(id)anObject forKey:(id)aKey cost:(int)aCost
+{
+    // Check if the key already exist
+    if ([_items objectForKey:aKey])
+        [self removeObjectForKey:aKey];
+
+    // Add object
+    [_items setObject:[_CPCacheItem cacheItemWithObject:anObject cost:aCost position:++_currentPosition] forKey:aKey];
+
+    // Clean cache to satisfy condition (< totalCostLimit & < countLimit) if necessary
+    [self _cleanCache];
+}
+
+/*!
+    Removes the object from the cache for the given key.
+    @param aKey the key of the object to be removed
+*/
+- (void)removeObjectForKey:(id)aKey
+{
+    // Call delegate method to warn that the object is going to be removed
+    if (_implementedDelegateMethods & CPCacheDelegate_cache_WillEvictObject)
+        [_delegate cache:self willEvictObject:[[_items objectForKey:aKey] object]];
+
+    [_items removeObjectForKey:aKey];
+}
+
+/*!
+    Removes all the objects from the cache.
+*/
+- (void)removeAllObjects
+{
+    _currentPosition = 0;
+
+    // Call delegate method to warn that the objects are going to be removed
+    if (_implementedDelegateMethods & CPCacheDelegate_cache_WillEvictObject)
+    {
+        var enumerator = [_items keyEnumerator],
+            value;
+
+        while (value = [enumerator nextObject])
+            [_delegate cache:self willEvictObject:[[_items objectForKey:value] object]];
+    }
+
+    [_items removeAllObjects];
+}
+
+
+#pragma mark Accessors
+
+/*!
+    Returns the count limit of the cache
+*/
+- (int)countLimit
+{
+    return _countLimit;
+}
+
+/*!
+    Sets the count limit of the cache.
+    Remove objects if not enough place to keep all of them
+    @param aCountLimit the new count limit
+*/
+- (void)setCountLimit:(int)aCountLimit
+{
+    _countLimit = aCountLimit;
+    [self _cleanCache];
+}
+
+/*!
+    Returns the total cost limit of the cache
+*/
+- (int)totalCostLimit
+{
+    return _totalCostLimit;
+}
+
+/*!
+    Sets the total cost limit of the cache.
+    Remove objects if not enough place to keep all of them
+    @param aTotalCostLimit the new total cost limit
+*/
+- (void)setTotalCostLimit:(int)aTotalCostLimit
+{
+    _totalCostLimit = aTotalCostLimit;
+    [self _cleanCache];
+}
+
+/*!
+    Returns the cache's delegate
+*/
+- (id)delegate
+{
+    return _delegate;
+}
+
+/*!
+    Sets the cache's delegate.
+    @param aDelegate the new delegate
+*/
+- (void)setDelegate:(id)aDelegate
+{
+    if (_delegate === aDelegate)
+        return;
+
+    _delegate = aDelegate;
+    _implementedDelegateMethods = 0;
+
+    if ([_delegate respondsToSelector:@selector(cache:willEvictObject:)])
+        _implementedDelegateMethods |= CPCacheDelegate_cache_WillEvictObject
+}
+
+
+#pragma mark Privates
+
+/*
+ * This method return the number of objects in the cache
+ */
+- (int)_count
+{
+    return [_items count];
+}
+
+/*
+ * This method return the total cost (addition of all object's cost in the cache)
+ */
+- (int)_totalCost
```

Comment:
what about some caching here?

- You have a variable `_totalCountCacheValid`
- in `- (int)_totalCost`
    - if `! _totalCountCacheValid`, you compute the value, cache it and return it
    - if  `_totalCountCacheValid`, you simply return the cached value
- when you add or remove an item, you simply mark the cache as invalid